server/rooms.js

```diff
@@ -1610,55 +1610,86 @@ let Rooms = Object.assign(getRoom, {
 	 * @param {AnyObject} options
 	 */
 	createBattle(formatid, options) {
-		const p1 = /** @type {User?} */ (options.p1);
-		const p2 = /** @type {User?} */ (options.p2);
-		if (p1 && p1 === p2) throw new Error(`Players can't battle themselves`);
-		if (p1) Ladders.cancelSearches(p1);
-		if (p2) Ladders.cancelSearches(p2);
+		/** @type {User?[]} */
+		const players = [options.p1, options.p2, options.p3, options.p4];
+		const gameType = Dex.getFormat(formatid).gameType;
+		if (gameType !== 'multi' && gameType !== 'free-for-all') {
+			if (players[2] || players[3]) {
+				throw new Error(`Four players were provided, but the format is a 2 player format.`);
+			} else {
+				delete players[2];
+				delete players[3];
+			}
+		}
+		for (const [i, p1] of players.entries()) {
+			if (!p1) continue;
+			for (const p2 of players.slice(i + 1)) {
+				if (p1 === p2) throw new Error(`Players can't battle themselves`);
+			}
+		}
+
+		for (const p of players) {
+			if (p) Ladders.cancelSearches(p);
+		}
 
 		if (Rooms.global.lockdown === true) {
-			if (p1) p1.popup("The server is restarting. Battles will be available again in a few minutes.");
-			if (p2) p2.popup("The server is restarting. Battles will be available again in a few minutes.");
+			for (const p of players) {
+				if (p) p.popup("The server is restarting. Battles will be available again in a few minutes.");
+			}
 			return;
 		}
 
-		// @ts-ignore
-		if (p1 && p2 && (p1.specialNextBattle || p2.specialNextBattle)) {
+		if (!players.includes(null)) {
 			// @ts-ignore
-			const p1special = p1.specialNextBattle, p2special = p2.specialNextBattle;
-			// @ts-ignore
-			p1.specialNextBattle = null;
-			// @ts-ignore
-			p2.specialNextBattle = null;
-
-			if (p1special !== p2special) {
-				p1.popup(`Your special battle settings don't match: "${p1special}" and "${p2special}"`);
-				p2.popup(`Your special battle settings don't match: "${p1special}" and "${p2special}"`);
+			const p1Special = players[0].specialNextBattle;
+			let mismatch = `"${p1Special}"`;
+			for (const p of players) {
+				// @ts-ignore
+				if (p && p.specialNextBattle !== p1Special) {
```

Comment:
```suggestion
				if (p.specialNextBattle !== p1Special) {
```

`p &&` is now no longer necessary.